io/disk/disktest.py

```diff
@@ -0,0 +1,135 @@
+            self.error('Gcc is needed for the test to be run')
+
+        self.disks = self.params.get('disks', default=self.workdir)
+        self.gigabytes = self.params.get('gigabytes', default=None)
+        self.chunk_mb = self.params.get('chunk_mb', default=None)
+        source = self.params.get('source', default='disktest.c')
+        makefile = self.params.get('make', default='Makefile')
+        c_file = os.path.join(self.datadir, source)
+        c_file_name = os.path.basename(c_file)
+        make_file = os.path.join(self.datadir, makefile)
+        make_file_name = os.path.basename(make_file)
+        dest_c_file = os.path.join(self.srcdir, c_file_name)
+        dest_m_file = os.path.join(self.srcdir, make_file_name)
+        shutil.copy(c_file, dest_c_file)
+        shutil.copy(make_file, dest_m_file)
+        self.memory_mb = memory.memtotal() / 1024 / 8
+        os.chdir(self.srcdir)
+        build.make(self.srcdir)
+
+    def one_disk_chunk(self, disks, chunk):
+        """
+        Tests one part of the disk by spawning a disktest instance.
+        :param disks: Directory (usually a mountpoint).
+        :param chunk: Portion of the disk used.
+        """
+        self.log.info("Testing %d MB files on %s in %d MB memory, chunk %s",
+                      self.chunk_mb, disks, self.memory_mb, chunk)
+        logfile = os.path.join(self.outputdir, "log.txt")
+        cmd = ("%s/disktest -m %d -f %s/testfile.%d -i -S >>%s 2>&1" %
+               (self.srcdir, self.chunk_mb, disks, chunk, logfile))
+        self.log.debug("Running '%s'", cmd)
+
+        proc = process.get_sub_process_klass(cmd)(cmd, shell=True)
+        proc.start()
+        proc.poll()
+        proc.wait()
+        return proc
+
+    def test(self):
+        """
+        Runs one iteration of disktest.
+
+        """
+        os.chdir(self.srcdir)
+        if self.chunk_mb is None:
+            self.chunk_mb = memory.memtotal() / 1024 / 8
+        if self.gigabytes is None:
+            free = 100  # cap it at 100GB by default
+            if not os.path.isdir(self.disks):
+                os.makedirs(self.disks)
+
+            free = min(disk.freespace(self.disks) / 1024 ** 3, free)
+            self.gigabytes = free
+            self.log.info("Resizing to %s GB", self.gigabytes)
+
+        if self.memory_mb > self.chunk_mb:
+            self.error("Too much RAM (%dMB) for this test to work" %
+                       self.memory_mb)
+
+        if self.chunk_mb == 0:
+            self.chunk_mb = 1
+        chunks = (1024 * self.gigabytes) / self.chunk_mb
+        self.log.info("Total of disk chunks that will be used: %s", chunks)
+        errors = []
+        for i in xrange(chunks):
+            proc = self.one_disk_chunk(self.disks, i)
+            retval = proc.wait()
+            if retval != 0:
+                errors.append(retval)
+        if errors:
+            self.fail("Errors from children: %s" % errors)
```

Comment:
Hello @Nosheen08, let's put the original test here:

```python
        os.chdir(self.disk_srcdir)
        if chunk_mb is None:
            chunk_mb = utils_memory.memtotal() / 1024 / 8
        if disks is None:
            disks = [self.tmpdir]
        if gigabytes is None:
            free = 100  # cap it at 100GB by default
            for disk in disks:
                free = min(utils.freespace(disk) / 1024 ** 3, free)
            gigabytes = free
            logging.info("Resizing to %s GB", gigabytes)
            sys.stdout.flush()

        self.chunk_mb = chunk_mb
        self.memory_mb = utils_memory.memtotal() / 1024 / 8
        if self.memory_mb > chunk_mb:
            raise error.TestError("Too much RAM (%dMB) for this test to work" %
                                  self.memory_mb)

        chunks = (1024 * gigabytes) / chunk_mb

        logging.info("Total of disk chunks that will be used: %s", chunks)
        for i in range(chunks):
            pids = []
            for disk in disks:
                pid = self.test_one_disk_chunk(disk, i)
                pids.append(pid)
            errors = []
            for pid in pids:
                (junk, retval) = os.waitpid(pid, 0)
                if (retval != 0):
                    errors.append(retval)
            if errors:
                raise error.TestError("Errors from children: %s" % errors)
```

compare to your version, there is one significant difference. The original version supports list of `disks`, for example `["/dev/sdb", "/dev/sdc", "/tmp/foo", "/tmp/bar"]`. Because of this the code is more complex to use the smallest size to fit in all disks, it supports running the `one_disk_chunk` in parallel and so on.

Your version keeps some of the complexity, but only allows single disk. In some cases that results in unnecessary code, which is acceptable if you intend to extend it in the future. But the main testing loop which spawns the `one_disk_chunk` is actually broken, because:

1. `one_disk_chunk` itself waits for command completion (I guess it's remnant from your debugging, because the main loop also waits for proc to finish).
2. You don't fail early, but append the `one_disk_chunk` test return value to errors and report them after all chunks are tested. That's a change from the original test, which failed with first broken chunk, not necessarily a mistake if it's really intended. The problem is, that the list of errors says nothing, because you don't know which chunk failed.

So I'd suggest porting the test as was originally implemented (with the support for multiple disks) and only adjust the bits related to avocado like `process`, exceptions and style. Alternatively you can say you're only interested in simple version, then feel free to implement it, but properly with reasonable error messages and without unnecessary code.